### PR TITLE
Add returnedObj expression support for apm-customize-enhance-plugin

### PR DESCRIPTION
**Pull request by chanjarster**
### add returnedObj expression support for apm-customize-enhance-plugin
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [x] Update the documentation to include this new feature.
- [x] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes https://github.com/apache/skywalking/issues/8292
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking-java/blob/main/CHANGES.md).

### CODE REVIEW

File: docs/en/setup/service-agent/java-agent/Customize-enhance-trace.md [hunk 1]
```diff
@@ -38,6 +38,15 @@ Implementing enhancements to custom classes requires two steps.
...
 	        </method>
+ 	        <method method="retString(java.lang.String)" operation_name="/retString" static="false">
+                <tag key="tag_ret">returnedObj</tag>
+                <log key="log_map">returnedObj</log>
+            </method>
+            <method method="retModel0(test.apache.skywalking.apm.testcase.customize.model.Model0)"
+              operation_name="/retModel0" static="false">
+                <tag key="tag_ret">returnedObj.model1.id</tag>
+                <log key="log_map">returnedObj.model1.getId()</log>
+            </method>
```
Comment (wu-sheng) [inline docs/en/setup/service-agent/java-agent/Customize-enhance-trace.md:49]: These lines seem not format correctly?

**Comment by wu-sheng** [comment]
One important question, what you changed is going to make the expressions invoked in both `before` and `after`? I feel it is.

File: docs/en/setup/service-agent/java-agent/Customize-enhance-trace.md (hunk 1 above)
Comment (chanjarster) [reply docs/en/setup/service-agent/java-agent/Customize-enhance-trace.md]: fixed

**Comment by chanjarster** [comment]
@wu-sheng Actually it doesn't. before method only evaluate arg expr, after method only evaluate returnedObj expr.

File: apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java [hunk 2]
```diff
@@ -54,33 +54,44 @@ void beforeMethod(Method method, Object[] allArguments) {
...
-            if (!spanTags.isEmpty()) {
-                for (Map.Entry<String, String> tag : spanTags.entrySet()) {
-                    span.tag(Tags.ofKey(tag.getKey()), tag.getValue());
-                }
-            }
-            if (!spanLogs.isEmpty()) {
-                span.log(System.currentTimeMillis(), spanLogs);
-            }
+            tagSpanTags(span, spanTags);
+            tagSpanLogs(span, spanLogs);
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java:64]: Aren't these tagging and logging things?

File: apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java [hunk 3]
```diff
@@ -54,33 +54,44 @@ void beforeMethod(Method method, Object[] allArguments) {
...
+
+        Map<String, String> tags = MethodConfiguration.getTags(configuration);
+        Map<String, String> spanTags = new HashMap<String, String>();
+        Map<String, String> logs = MethodConfiguration.getLogs(configuration);
+        Map<String, String> spanLogs = new HashMap<String, String>();
+
+        evalReturnAndPopulate(context, tags, spanTags);
+        evalReturnAndPopulate(context, logs, spanLogs);
+
+        final AbstractSpan localSpan = ContextManager.activeSpan();
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java:89]: All variables here(configuration, context, tags, spanTags, localSpan) could share with before method by using plugin V2 APIs(https://skywalking.apache.org/docs/skywalking-java/v8.8.0/en/setup/service-agent/java-agent/java-plugin-development-guide/#v2-apis-1)

**MethodInvocationContext context** is the key point to this. You could save a lot resources and GC cost by sharing this.

**Comment by wu-sheng** [comment]
https://github.com/apache/skywalking-java/blob/a65f6bd85457356dd2279adc26322aa7005d19d1/apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java#L39-L42

This is not your fault, but I hope you could help on polishing this. The `if/else` doesn't mean `create local span or not`(they both did), but more about whether to create more custom tags/logs. Could you make this logic more clear when adopting V2 APIs?

File: apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java (hunk 3 above)
Comment (chanjarster) [reply apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java]: Updated

File: apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/define/CustomizeInstanceInstrumentation.java [hunk 4]
```diff
@@ -71,7 +71,7 @@ public boolean isOverrideArgs() {
         }
     }
 
-    @Override
+  @Override
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/define/CustomizeInstanceInstrumentation.java:74]: Mis-format?

File: apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java [hunk 5]
```diff
@@ -18,69 +18,118 @@
...
+        evalReturnAndPopulate(evalContext, tags, spanTags);
+        evalReturnAndPopulate(evalContext, logs, spanLogs);
+
+        tagSpanTags(localSpan, spanTags);
+        tagSpanLogs(localSpan, spanLogs);
+
+        spanTags.clear();
+        spanLogs.clear();
+
+        ContextManager.stopSpan(localSpan);
```
Comment (wu-sheng) [inline apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java:132]: Please put `stopSpan` in the `try/final` block.

Comment (wu-sheng) [reply apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java:132]: Then in the `handleMethodException `, `if (ContextManager.isActive()) {` is not required anymore. And the span/context would not be leaked.

**Comment by wu-sheng** [comment]
Generally good, add some polish comments.

File: apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/define/CustomizeInstanceInstrumentation.java (hunk 4 above)
Comment (chanjarster) [reply apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/define/CustomizeInstanceInstrumentation.java]: fixed

File: apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java (hunk 5 above)
Comment (chanjarster) [reply apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java]: changed

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**